File: _posts/2023-07-09-benchmarking-recursion-in-different-languages/assets/fibonacci.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature 'say';

sub fibonacci {
    my $term = shift;
    # Base case.
    if ($term <= 1) {
        return $term;
    }
    # Recursive step.
    return fibonacci($term - 1) + fibonacci($term - 2);
}

sub main {
    if (scalar @ARGV >= 1) {
        my $value = fibonacci(int($ARGV[0]));
        say($value);
    }
}

unless (caller) {
    main();
}

__END__

=head1 DESCRIPTION

Calculate a Fibonacci sequence number from user input.

=cut

=head1 FUNCTIONS

=over

=item fibonacci($term)

Calculate a Fibonacci sequence number.

$term C<int> is the sequence number to produce. Don't use large numbers as the
time complexity is exponential.

Returns C<int>, the sequence number.

=item main

Calculate a Fibonacci sequence number from user input.

=back

=cut

File: _posts/2023-07-09-benchmarking-recursion-in-different-languages/assets/fibonacci.dart
/// Calculate a Fibonacci sequence number from user input.
void main(List<String> args) {
    final term = int.parse(args[0]);
    print(fibonacci(term));
}

/// Calculate the Fibonacci sequence number for the [n]-th term.
///
/// Don't use large numbers as the time complexity is exponential.
int fibonacci(int n) {
  // Base case: the first two numbers in the sequence are 0 and 1.
  if (n <= 1) {
    return n;
  }

  // Recursive step: The nth number in the sequence is the sum of the (n - 1)th
  // and (n - 2)th numbers.
  return fibonacci(n - 1) + fibonacci(n - 2);
}
